### PR TITLE
Exploration: Retry STATUS_UNSUCCESSFUL from CNG

### PR DESCRIPTION
**Pull request by bartonjs**
This is an attempt to address #37994, but since it's a very intermittent failure it's not really easy to say if the fix "works" or not.

Success will be measured by seeing no failures in CI for a month or so.  If it doesn't work, we'll just back it out.

### CODE REVIEW

File: src/Common/src/Internal/Cryptography/CngCommon.SignVerify.cs [hunk 1]
```diff
@@ -19,11 +21,23 @@ public static unsafe byte[] SignHash(this SafeNCryptKeyHandle keyHandle, ReadOnl
...
+                errorCode = Interop.NCrypt.NCryptSignHash(keyHandle, pPaddingInfo, hash, hash.Length, signature, signature.Length, out numBytesNeeded, paddingMode);
+            }
+
             if (errorCode == ErrorCode.NTE_BUFFER_TOO_SMALL)
             {
                 signature = new byte[numBytesNeeded];
                 errorCode = Interop.NCrypt.NCryptSignHash(keyHandle, pPaddingInfo, hash, hash.Length, signature, signature.Length, out numBytesNeeded, paddingMode);
             }
+
+            if (errorCode == ErrorCode.STATUS_UNSUCCESSFUL)
```
Comment (krwq) [inline src/Common/src/Internal/Cryptography/CngCommon.SignVerify.cs]: this might actually be third attempt, right? I'm fine either way

File: src/Common/src/System/Security/Cryptography/RSACng.EncryptDecrypt.cs [hunk 2]
```diff
@@ -224,47 +224,92 @@ private unsafe byte[] EncryptOrDecrypt(SafeNCryptKeyHandle key, ReadOnlySpan<byt
...
             }
 
-            Array.Resize(ref output, numBytesNeeded);
+            if (numBytesNeeded != output.Length)
+            {
+                Span<byte> writtenSpan = output.AsSpan(0, numBytesNeeded);
+                byte[] ret = writtenSpan.ToArray();
+                CryptographicOperations.ZeroMemory(writtenSpan);
+                output = ret;
+            }
```
Comment (stephentoub) [inline src/Common/src/System/Security/Cryptography/RSACng.EncryptDecrypt.cs]: This doesn't look related to this fix.  You're doing this to be able to clear the original buffer?

File: src/Common/src/Internal/Cryptography/CngCommon.SignVerify.cs (hunk 1 above)
Comment (bartonjs) [reply src/Common/src/Internal/Cryptography/CngCommon.SignVerify.cs]: It could be call 1: Unsuccessful, 2: too small, 3: Unsuccessful (large enough array), 4: last attempt, yeah.

But I'm pretty sure that this one uses the public keysize internally to understand too small, and all the failures were related to the private key.  So it's more likely (in retail): 1: Success, or 1: Unsuccessful, 2: last attempt, since we use a big enough buffer.

File: src/Common/src/System/Security/Cryptography/RSACng.EncryptDecrypt.cs (hunk 2 above)
Comment (bartonjs) [reply src/Common/src/System/Security/Cryptography/RSACng.EncryptDecrypt.cs]: Yeah, it was a "grr, this is wrong and in a method I'm fixing".  It looks like it broke UWP somehow, so I'm going to undo this part and just do it as a separate change.

**Comment by ViktorHofer** [comment]
Does this affect UAP as well? This just failed again on it: https://dnceng.visualstudio.com/public/_build/results?buildId=366753&view=ms.vss-test-web.build-test-results-tab&runId=11095708&resultId=147147&paneView=debug
